src/gtest/test_noteencryption.cpp

```diff
@@ -0,0 +1,170 @@
+    }
+
+    for (int i = 0; i < 255; i++) {
+        auto ciphertext = b.encrypt(pk_enc, message);
+
+        {
+            ZCNoteDecryption decrypter(sk_enc);
+
+            // Test decryption
+            auto plaintext = decrypter.decrypt(ciphertext, b.get_epk(), uint256(), i);
+            ASSERT_TRUE(plaintext == message);
+
+            // Test wrong nonce
+            ASSERT_THROW(decrypter.decrypt(ciphertext, b.get_epk(), uint256(), (i == 0) ? 1 : (i - 1)), std::runtime_error);
+        
+            // Test wrong ephemeral key
+            ASSERT_THROW(decrypter.decrypt(ciphertext, ZCNoteEncryption::generate_privkey(uint256()), uint256(), i), std::runtime_error);
+        
+            // Test wrong seed
+            ASSERT_THROW(decrypter.decrypt(ciphertext, b.get_epk(), uint256S("11035d60bc1983e37950ce4803418a8fb33ea68d5b937ca382ecbae7564d6a77"), i), std::runtime_error);
+        
+            // Test corrupted ciphertext
+            ciphertext[10] ^= 0xff;
+            ASSERT_THROW(decrypter.decrypt(ciphertext, b.get_epk(), uint256(), i), std::runtime_error);
+            ciphertext[10] ^= 0xff;
+        }
+
+        {
+            // Test wrong private key
+            uint256 sk_enc_2 = ZCNoteEncryption::generate_privkey(uint256());
+            ZCNoteDecryption decrypter(sk_enc_2);
+
+            ASSERT_THROW(decrypter.decrypt(ciphertext, b.get_epk(), uint256(), i), std::runtime_error);
+        }
+
+        {
+            TestNoteDecryption decrypter(sk_enc);
+
+            // Test decryption
+            auto plaintext = decrypter.decrypt(ciphertext, b.get_epk(), uint256(), i);
+            ASSERT_TRUE(plaintext == message);
+
+            // Test wrong public key (test of KDF)
+            decrypter.change_pk_enc(uint256());
+            ASSERT_THROW(decrypter.decrypt(ciphertext, b.get_epk(), uint256(), i), std::runtime_error);
+        }
+    }
+
+    // Nonce space should run out here
+    ASSERT_THROW(b.encrypt(pk_enc, message), std::logic_error);
+}
+
+uint256 test_prf(
+    unsigned char distinguisher,
+    uint256 x,
+    uint256 y
+) {
+    *x.begin() &= 0x0f;
+    *x.begin() |= distinguisher;
+    CSHA256 hasher;
+    hasher.Write(x.begin(), 32);
+    hasher.Write(y.begin(), 32);
+
+    uint256 ret;
+    hasher.FinalizeNoPadding(ret.begin());
+    return ret;
+}
+
+TEST(noteencryption, prf_addr)
+{
+    for (size_t i = 0; i < 100; i++) {
+        uint256 a_sk = libzcash::random_uint256();
+        uint256 rest;
+        ASSERT_TRUE(
+            test_prf(0xc0, a_sk, rest) == PRF_addr_a_pk(a_sk)
+        );
+    }
+
+    for (size_t i = 0; i < 100; i++) {
+        uint256 a_sk = libzcash::random_uint256();
+        uint256 rest;
+        *rest.begin() = 0x01;
+        ASSERT_TRUE(
+            test_prf(0xc0, a_sk, rest) == PRF_addr_sk_enc(a_sk)
+        );
+    }
+}
+
+TEST(noteencryption, prf_nf)
+{
+    for (size_t i = 0; i < 100; i++) {
+        uint256 a_sk = libzcash::random_uint256();
```

Comment:
We should have some sort of a method for generating `a_sk`'s that zeroes out the leading 4 bits (not sure if that's in the scope of this PR).